lib/galaxy/managers/workflows.py

```diff
@@ -413,9 +413,21 @@ def cancel_invocation(self, trans, decoded_invocation_id):
             trans.sa_session.add(workflow_invocation)
             with transaction(trans.sa_session):
                 trans.sa_session.commit()
-        else:
-            # TODO: More specific exception?
-            raise exceptions.MessageException("Cannot cancel an inactive workflow invocation.")
+
+        for step in workflow_invocation.steps:
+            for job in step.jobs:
+                job.mark_deleted()
```

Comment:
For consistency with `JobManager.stop()`:

```suggestion
                job.mark_deleted(self.app.config.track_jobs_in_database)
```

Reply:
Actually, it would probably be better to use `galaxy.managers.jobs.JobManager.stop()`:

```python
from galaxy.managers.jobs import JobManager

...
job_manager = JobManager(trans.app)
...

                job_manager.stop(flush=False)
```

with this additional patch:

```diff
diff --git a/lib/galaxy/managers/jobs.py b/lib/galaxy/managers/jobs.py
index 54c01bcea4..18589609c8 100644
--- a/lib/galaxy/managers/jobs.py
+++ b/lib/galaxy/managers/jobs.py
@@ -239,12 +239,13 @@ class JobManager:
         trans.sa_session.refresh(job)
         return job
 
-    def stop(self, job, message=None):
+    def stop(self, job, message=None, flush=True):
         if not job.finished:
             job.mark_deleted(self.app.config.track_jobs_in_database)
-            session = self.app.model.session
-            with transaction(session):
-                session.commit()
+            if flush:
+                session = self.app.model.session
+                with transaction(session):
+                    session.commit()
             self.app.job_manager.stop(job, message=message)
             return True
         else:
diff --git a/lib/galaxy/structured_app.py b/lib/galaxy/structured_app.py
index 22f03c10e1..c83718a2d7 100644
--- a/lib/galaxy/structured_app.py
+++ b/lib/galaxy/structured_app.py
@@ -41,6 +41,7 @@ from galaxy.workflow.trs_proxy import TrsProxy
 if TYPE_CHECKING:
     from galaxy.config_watchers import ConfigWatchers
     from galaxy.jobs import JobConfiguration
+    from galaxy.jobs.manager import JobManager
     from galaxy.managers.collections import DatasetCollectionManager
     from galaxy.managers.hdas import HDAManager
     from galaxy.managers.histories import HistoryManager
@@ -109,7 +110,7 @@ class MinimalManagerApp(MinimalApp):
     installed_repository_manager: "InstalledRepositoryManager"
     user_manager: Any
     job_config: "JobConfiguration"
-    job_manager: Any  # galaxy.jobs.manager.JobManager
+    job_manager: "JobManager"
     job_metrics: JobMetrics
     dynamic_tool_manager: Any  # 'galaxy.managers.tools.DynamicToolManager'
     genomes: "Genomes"

```

BTW, it is very confusing that we have 2 `JobManager` classes and `app.job_manager` is not an instance of `galaxy.managers.jobs.JobManager` !

Reply:
I think the conditional flush in the diff is mandatory since the JobHandlerStopQueue receives the job by id, and I believe the queue itself uses a different thread and should have a different sqlalchemy session that doesn't see the changes to the job state yet.

My understanding of how this currently works is that the deleted state gets picked up by the monitor without needing to call `self.app.job_manager.stop(job, message=message)`. (And maybe existing code should be refactored to not call `app.job_manager.stop` ?)